### ISSUE TITLE
API to access link vertices missing?

### ISSUE THREAD

**Issue by jbaayen** [OpenWaterAnalytics/EPANET#151]
EPANET 2.0 allows one to add vertices to links. When exporting a network, these are then listed in the "VERTICES" section of the .INP file.

However, in 2.1, I do not see API to access this vertex data.

Is this intentional? If not, is vertex API scheduled to be introduced in the future?

**Comment by eladsal** [comment]
The vertices were never accessible via the Toolkit's APIs. It is planned and discussed here #46

**Comment by michaeltryby** [comment]
Hey @jbaayen could you describe your use case for us? You want to add vertices to links,  programmatically modifying the data model? Is the intent to have them appear in the input file?

**Comment by jbaayen** [comment]
Actually, the other way around: I'd like to read the input file using the API, and create a visualisation of the network.

**Comment by bradleyjeck** [comment]
@jbaayen if you are using R, the [epanetReader](https://github.com/bradleyjeck/epanetReader) package might be of use.  The package reads the inp file, including vertices.

**Comment by jbaayen** [comment]
@bradleyjeck thanks for the pointer!
We're using Python here though. A separate INP file parser could certainly be an option, it just seems a shame to have two parallel parser implementations parse the same file.

**Comment by bemcdonnell** [comment]
@jbaayen, we’ve had this debate on the SWMM side of OWA a few times. We don’t have plans to add this to our API since the engine itself does not use coordinates ( or TAGS) in any way. 

Imagine a future without us needing standard input files for SWMM or EPANET. Imagine just a set of functions allowing you to initialize a model from whatever format you want - say GeoJSON.  Network visualization rendering data would come from the GeoJSON file and would not need to be plucked through an API.  

IMO, including coordinate support into the API seems like we’re committing ourselves to the traditional input file for the rest of time. What are your thoughts?

**Comment by dickinsonre** [comment]
However, the coordinates in SWMM at least could be used have a better estimate of manhole losses based on the geometry of the incoming and outgoing links.  It is a big lack of engine power in SWMM 4/5 in my opinion.  I do like the concept of using other formats for the inp file but still think using the inp file as a core is hyper important.

**Comment by michaeltryby** [comment]
@dickinsonre The problem with the input file is that the parser has been welded onto the data model within the engine. @bemcdonnell and I aren't proposing to get rid of the input file, we are proposing to decouple the parser from the data model to allow other modes of interacting with it. Make sense?

**Comment by dickinsonre** [comment]
Yes, that makes perfect sense @michaeltryby and @bemcdonnell Thanks also for making this a very good and active GitHub site.

**Comment by samhatchett** [comment]
Agreed with @michaeltryby - this is the direction epanet is headed as well. With a reference C implementation parser that probably loads an intermediate data model with parsed `inp` relational data, and then can build a hydraulic model or offer that data to other consumers. 

As another example, we use epanet coordinates to reproject to lon/lat and then integrate with the national elevation dataset to ensure that junction elevations are not wildly wrong.

**Comment by bouziot** [comment]
@jbaayen @samhatchett a Python EPANET parser is being in development, as it was needed for a large project with messy networks. It's able to parse EPANET files (.inp, .msx, .rpt) to pandas dataframes, but also rename nodes and links and compose the updates back to an inp file. It's still under heavy development and testing but will likely be operational in a beta phase end of May or June. I will be very pleased to upload it as part of the OWA repo once the initial tests are completed as a beta version (so that we can also improve it together).

**Comment by samhatchett** [comment]
@bouziot thanks for commenting. WNTR also has a module that does what you describe - can you say more about how your development might be different, or what the relative merits are of using INP as the intermediate exchange format versus having a single authoritative data model?

**Comment by eldemet** [comment]
1. We also resolved this by writing an external parser in MATLAB (see function [NodeCoords](https://github.com/OpenWaterAnalytics/EPANET-Matlab-Toolkit/blob/ab0cb3604bfb4f572a01014c7a56d66b8d905c68/epanet_matlab_toolkit/epanet.m#L11295) in `EPANET MATLAB Toolkit`.  (Actually we also parse the whole inp/msx/bin file(s) and translate it into a structure/dataframe.) 

2. For the library, as discussed in [#46](https://github.com/OpenWaterAnalytics/EPANET/issues/46) I think we could provide a function in EPANET for returning/parsing the additional information stored, mainly for the vertices. Currently a lot of people are resolving it using parsers, which is not optimal. Also, we should keep in mind that at some point someone might want to study a new research challenge in EPANET modelling we haven't considered yet which require this information, so it's not just a visualization issue.

3. On the other hand, I overall I agree with @bemcdonnell . I believe GeoJSON could be nicely integrated with EPANET for modelling the topology. This is already with case with some of the QGIS plugins we develop at KIOS, and this is also an export feature of the new EPANET GUI (MTP4r2).  So we already have a standardization of an EPANET-compatible GeoJSON structure, which we can build upon. I am under the impression that different teams are working on this. Perhaps we should all collaborate to agree on a common vocabulary and structure, to avoid re-inventing the wheel? We could event take this one step further and standardize a 1:1 relation between the INP files and GeoJSON (however this would need to consider adding at least demand time series and control rules into the JSON).

**Comment by bouziot** [comment]
@samhatchett I think the content of issues such as the one we talk in requires, at least partially and in the present versions of EPANET, a form of text parsing in order to obtain all information needed from a saved topology. @eldemet  also explained that the MATLAB toolkit partially parses information externally. While the standardization of a new structure that you discuss would be ideal, presently most teams work with the EPANET I/O text files to communicate their work flows and product; the text file formats themselves, while not perhaps the optimal standard, provide platform and version-independent background to understand and communicate different projects. Any other parsing projects I see provide this functionality based on the lower-level API, therefore introducing - up to a point - version and platform dependency.

**Comment by kaklise** [comment]
I agree that INP parser implementations are a sizable undertaking, there are several available, and there are many edge cases to consider.  I also agree that there is a need to support the INP file format and for a flexible API to build models from scratch.  WNTR includes options for both, and can read/write Vertices using INP files and modify/add Vertices using the API.  Additional use cases for Vertices could be added within the Python framework.

**Comment by bouziot** [comment]
I'll gladly share information on the actual code and use once I've ensured that major bugs are out of the way, otherwise it will be probably a loss of your (more precious) code evaluation time. Now that I know WNTR I'll also check its parsing module, and perhaps fit it in my use case as well. 

The C parser is indeed a very good idea and it will have unmatched speed. I saw some issues regarding cross-platform behavior and dependence on how the dll is compiled [in the ctypes implementation](http://github.com/OpenWaterAnalytics/epanet-python/tree/dev/epanet-module) but if I'm not mistaken SWIG bypasses these issues. I also don't know how well regexp can be handled in C (so that we can query and conditionally change names in nodes/links), but then I'm not very knowledgeable in C overall. In any case, I could perhaps help in the team wrapping such an endeavor to Python.  

P.S. @samhatchett I'll be happy to update my details and generally become more active in the community - give me a few days. By the way, is there perhaps a topic more relevant to discuss about this?
